File: kotlin/src/main/kotlin/models/MediaContent.kt
package aethermesh.media.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Immutable description of a single piece of media stored on the Aether network.
 * Primary key is [contentHash] — SHA-256 hex of the raw encoded bytes.
 */
@Serializable
data class MediaContent(
    @SerialName("content_hash")
    val contentHash: String,
    val title: String,
    @SerialName("duration_ms")
    val durationMs: Long,
    val codec: String,
    @SerialName("content_type")
    val contentType: String,
    @SerialName("creator_uhid")
    val creatorUhid: String,
    @SerialName("size_bytes")
    val sizeBytes: Long,
    val tags: List<String> = emptyList(),
    @SerialName("thumbnail_hash")
    val thumbnailHash: String? = null,
) {
    /**
     * Human-readable duration:
     * - 0 ms → "Live"
     * - < 1 hour → "M:SS"
     * - >= 1 hour → "H:MM:SS"
     */
    val formattedDuration: String
        get() = when {
            durationMs <= 0L -> "Live"
            durationMs < 3_600_000L -> {
                val totalSec = durationMs / 1000L
                val minutes = totalSec / 60L
                val seconds = totalSec % 60L
                "$minutes:${seconds.toString().padStart(2, '0')}"
            }
            else -> {
                val totalSec = durationMs / 1000L
                val hours   = totalSec / 3600L
                val minutes = (totalSec % 3600L) / 60L
                val seconds = totalSec % 60L
                "$hours:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}"
            }
        }

    val isVideo: Boolean get() = contentType.lowercase().startsWith("video/")
    val isAudio: Boolean get() = contentType.lowercase().startsWith("audio/")
}

// ── MediaReactionType ──────────────────────────────────────────────────────────

enum class MediaReactionType(val value: Int) {
    LIKE(1),
    SHARE(2),
    COMMENT(3),
    SUPER_REACT(4);

    companion object {
        fun fromValue(v: Int): MediaReactionType =
            entries.firstOrNull { it.value == v }
                ?: throw IllegalArgumentException("Unknown MediaReactionType value: $v")
    }
}

// ── MediaReaction ──────────────────────────────────────────────────────────────

/**
 * A timestamped reaction sent by a viewer.
 *
 * Validation rules:
 * - [message] is required (non-blank) for [MediaReactionType.COMMENT].
 * - [message] must be null for all other types.
 */
@Serializable
data class MediaReaction(
    @SerialName("reaction_id")
    val reactionId: String,
    @SerialName("content_hash")
    val contentHash: String,
    @SerialName("from_uhid")
    val fromUhid: String,
    val type: MediaReactionType,
    @SerialName("position_ms")
    val positionMs: Long,
    val message: String?,
    @SerialName("sent_at_ms")
    val sentAtMs: Long,
) {
    init {
        require(contentHash.isNotBlank()) { "contentHash must not be empty" }
        require(fromUhid.isNotBlank())    { "fromUhid must not be empty" }
        require(positionMs >= 0L)         { "positionMs must be >= 0" }
        if (type == MediaReactionType.COMMENT) {
            require(!message.isNullOrBlank()) {
                "A message is required for COMMENT reactions"
            }
        } else {
            require(message == null) {
                "message must be null for ${type.name} reactions"
            }
        }
    }
}

// ── MediaProfile ───────────────────────────────────────────────────────────────

@Serializable
data class MediaProfile(
    val uhid: String,
    @SerialName("display_name")
    val displayName: String,
    @SerialName("avatar_hash")
    val avatarHash: String?,
    val bio: String?,
    @SerialName("aethermesh_tag")
    val aethermeshTagValue: String,
    @SerialName("follower_count")
    val followerCount: Int,
    @SerialName("following_count")
    val followingCount: Int,
    @SerialName("content_count")
    val contentCount: Int,
    @SerialName("is_verified")
    val isVerified: Boolean,
    @SerialName("joined_at_ms")
    val joinedAtMs: Long,
) {
    private companion object { const val SHORT_BIO_MAX = 120 }

    /**
     * Bio trimmed to 120 characters at the last word boundary, with "…" appended.
     * Returns "" when [bio] is null or blank.
     */
    val shortBio: String
        get() {
            val trimmed = bio?.trim() ?: return ""
            if (trimmed.isEmpty()) return ""
            if (trimmed.length <= SHORT_BIO_MAX) return trimmed
            val cut = trimmed.substring(0, SHORT_BIO_MAX)
            val lastSpace = cut.lastIndexOf(' ')
            val boundary = if (lastSpace > 0) lastSpace else SHORT_BIO_MAX
            return cut.substring(0, boundary).trimEnd() + "…"
        }
}

// ── LiveStream ─────────────────────────────────────────────────────────────────

@Serializable
data class LiveStream(
    @SerialName("stream_id")
    val streamId: String,
    val title: String,
    @SerialName("creator_uhid")
    val creatorUhid: String,
    val codec: String,
    @SerialName("segment_duration_ms")
    val segmentDurationMs: Int,
    @SerialName("started_at_ms")
    val startedAtMs: Long,
    @SerialName("viewer_count")
    val viewerCount: Int,
    @SerialName("is_active")
    val isActive: Boolean,
    val tags: List<String> = emptyList(),
) {
    /** Wall-clock milliseconds since the broadcast started.  Always >= 0. */
    val elapsedMs: Long
        get() = maxOf(0L, System.currentTimeMillis() - startedAtMs)

    val elapsedFormatted: String
        get() {
            val totalSec = elapsedMs / 1000L
            val hours   = totalSec / 3600L
            val minutes = (totalSec % 3600L) / 60L
            val seconds = totalSec % 60L
            return if (hours > 0L) {
                "$hours:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}"
            } else {
                "$minutes:${seconds.toString().padStart(2, '0')}"
            }
        }
}

// ── MediaFeedItem ──────────────────────────────────────────────────────────────

@Serializable
data class MediaFeedItem(
    val content: MediaContent,
    @SerialName("like_count")
    val likeCount: Int,
    @SerialName("share_count")
    val shareCount: Int,
    @SerialName("comment_count")
    val commentCount: Int,
    @SerialName("watch_count")
    val watchCount: Int,
    @SerialName("is_live")
    val isLive: Boolean,
    @SerialName("stream_id")
    val streamId: String?,
    @SerialName("top_reactions")
    val topReactions: List<MediaReaction> = emptyList(),
    @SerialName("published_at_ms")
    val publishedAtMs: Long,
    @SerialName("watched_ms")
    val watchedMs: Long = 0L,
) {
    val isNew: Boolean
        get() = (System.currentTimeMillis() - publishedAtMs) < 86_400_000L

    val reactionTotal: Int
        get() = likeCount + shareCount + commentCount
}

File: typescript/src/models/MediaContent.ts
export interface MediaContent {
  contentHash: string;
  title: string;
  durationMs: number;
  codec: string;
  contentType: string;
  creatorUhid: string;
  sizeBytes: number;
  createdAtMs: number;
  thumbnailHash: string | null;
  tags: string[];
}

// ── Wire format ───────────────────────────────────────────────────────────────

/** Snake_case JSON representation of MediaContent as sent/received on the wire. */
export interface MediaContentWire {
  content_hash: string;
  title: string;
  duration_ms: number;
  codec: string;
  content_type: string;
  creator_uhid: string;
  size_bytes: number;
  created_at_ms: number;
  thumbnail_hash: string | null;
  tags: string[];
}

export function toWire(content: MediaContent): MediaContentWire {
  return {
    content_hash:   content.contentHash,
    title:          content.title,
    duration_ms:    content.durationMs,
    codec:          content.codec,
    content_type:   content.contentType,
    creator_uhid:   content.creatorUhid,
    size_bytes:     content.sizeBytes,
    created_at_ms:  content.createdAtMs,
    thumbnail_hash: content.thumbnailHash,
    tags:           content.tags,
  };
}

export function fromWire(obj: MediaContentWire): MediaContent {
  return {
    contentHash:   obj.content_hash,
    title:         obj.title,
    durationMs:    obj.duration_ms,
    codec:         obj.codec,
    contentType:   obj.content_type,
    creatorUhid:   obj.creator_uhid,
    sizeBytes:     obj.size_bytes,
    createdAtMs:   obj.created_at_ms,
    thumbnailHash: obj.thumbnail_hash,
    tags:          obj.tags,
  };
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/**
 * Returns a human-readable duration string.
 *
 * - 0  → "Live"
 * - < 3 600 000 ms → "M:SS" or "MM:SS"  (e.g. "4:32")
 * - >= 3 600 000 ms → "H:MM:SS"          (e.g. "1:23:45")
 */
export function formattedDuration(durationMs: number): string {
  if (durationMs <= 0) return "Live";

  const totalSeconds = Math.floor(durationMs / 1000);
  const hours   = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);
  const seconds = totalSeconds % 60;

  const ss = seconds.toString().padStart(2, "0");

  if (hours > 0) {
    const mm = minutes.toString().padStart(2, "0");
    return `${hours}:${mm}:${ss}`;
  }

  return `${minutes}:${ss}`;
}

export function isVideo(content: MediaContent): boolean {
  return content.contentType.toLowerCase().startsWith("video/");
}

export function isAudio(content: MediaContent): boolean {
  return content.contentType.toLowerCase().startsWith("audio/");
}
